frappe/desk/form/linked_with.py

```diff
@@ -1,14 +1,119 @@
+from frappe import _
 from frappe.model.meta import is_single
 from frappe.modules import load_doctype_module
-import frappe.desk.form.meta
-import frappe.desk.form.load
-from six import string_types
-from collections import defaultdict
+
+
+@frappe.whitelist()
+def get_submitted_linked_docs(doctype, name, docs=None):
+	"""
+	Get all nested submitted linked doctype linkinfo
+
+	Arguments:
+		doctype (str) - The doctype for which get all linked doctypes
+		name (str) - The docname for which get all linked doctypes
+
+	Keyword Arguments:
+		docs (list of dict) - (Optional) Get list of dictionary for linked doctype.
+
+	Returns:
+		dict - Return list of documents and link count
+	"""
+
+	if not docs:
+		docs = []
+
+	linkinfo = get_linked_doctypes(doctype)
+	linked_docs = get_linked_docs(doctype, name, linkinfo)
+
+	link_count = 0
+	for link_doctype, link_names in linked_docs.items():
+		for link in link_names:
+			docinfo = link.update({"doctype": link_doctype})
+			validated_doc = validate_linked_doc(docinfo)
+
+			if not validated_doc:
+				continue
+
+			link_count += 1
+			if link.name in [doc.get("name") for doc in docs]:
+				continue
+
+			links = get_submitted_linked_docs(link_doctype, link.name, docs)
+			docs.append({
+				"doctype": link_doctype,
+				"name": link.name,
+				"docstatus": link.docstatus,
+				"link_count": links.get("count")
+			})
+
+	# sort linked documents by ascending number of links
+	docs.sort(key=lambda doc: doc.get("link_count"))
+	return {
+		"docs": docs,
+		"count": link_count
+	}
+
+
+@frappe.whitelist()
+def cancel_all_linked_docs(docs):
+	"""
+	Cancel all linked doctype
+
+	Arguments:
+		docs (str) - It contains all list of dictionaries of a linked documents.
+	"""
+
+	docs = json.loads(docs)
+	for i, doc in enumerate(docs, 1):
+		if validate_linked_doc(doc) is True:
+			frappe.publish_progress(percent=i * 100 / len(docs), title=_("Cancelling documents"))
+			linked_doc = frappe.get_doc(doc.get("doctype"), doc.get("name"))
+			linked_doc.cancel()
+
+
+def validate_linked_doc(docinfo):
+	"""
+	Validate a document to be submitted and non-exempted from auto-cancel.
+
+	Args:
+		docs (dict): The document to check for submitted and non-exempt from auto-cancel
+
+	Returns:
+		bool: True if linked document passes all validations, else False
+	"""
+
+	# skip non-submittable doctypes since they don't need to be cancelled
+	if not frappe.get_meta(docinfo.get('doctype')).is_submittable:
+		return False
+
+	# skip draft or cancelled documents
+	if docinfo.get('docstatus') != 1:
+		return False
+
+	# skip other doctypes since they don't need to be cancelled
+	auto_cancel_exempt_doctypes = get_exempted_doctypes()
+	if docinfo.get('doctype') in auto_cancel_exempt_doctypes:
+		return False
+
+	return True
+
+
+def get_exempted_doctypes():
+	""" Get list of doctypes exempted from being auto-cancelled """
+
+	auto_cancel_exempt_doctypes = []
+	for doctypes in frappe.get_hooks('auto_cancel_exempt_doctypes'):
```

Comment:
@vishdha Please rename the hook as `auto_cancel_exempted_doctypes`

And update the docs accordingly. Sorry for the delayed review on this.